src/library/scala/collection/immutable/Vector.scala

```diff
@@ -526,6 +565,24 @@ final class Vector[+A] private[immutable] (private[collection] val startIndex: I
     s
   }
 
+  override def equals(o: Any): Boolean = o match {
+    case that: Vector[_] =>
+      if (this eq that) true
+      else if (this.length != that.length) false
+      else if ( //
+        this.startIndex == that.startIndex && //
+          this.endIndex == that.endIndex && //
+          (this.display0 eq that.display0) && //
+          (this.display1 eq that.display1) && //
+          (this.display2 eq that.display2) && //
+          (this.display3 eq that.display3) && //
+          (this.display4 eq that.display4) && //
+          (this.display5 eq that.display5) //
+      ) true
```

Comment:
In which situations would this apply where `this eq that` doesn't also?

Reply:
We can get here if we reuse a vector builder.

```
scala> val b = collection.immutable.Vector.newBuilder[Any]
b: scala.collection.mutable.Builder[Any,scala.collection.immutable.Vector[Any]] = scala.collection.immutable.VectorBuilder@2e5b7fba

scala> b += null
res0: b.type = scala.collection.immutable.VectorBuilder@2e5b7fba

scala> val r1, r2 = b.result()
r1: scala.collection.immutable.Vector[Any] = Vector(null)
r2: scala.collection.immutable.Vector[Any] = Vector(null)

scala> assert(r1 ne r2)

scala> def show[A](v: Vector[A]) = (0 to 5).map(i => "display" + i).foreach(f => {val fld = classOf[Vector[_]].getDeclaredField(f); fld.setAccessible(true); println(Option(fld.get(v)).map(_.asInstanceOf[Array[Object]]))})
show: [A](v: Vector[A])Unit

scala> show(r1)
Some([Ljava.lang.Object;@29f95272)
None
None
None
None
None

scala> show(r2)
Some([Ljava.lang.Object;@29f95272)
None
None
None
None
None
```

Perhaps rare, but the check will cheaply fail in the common case so I'd say we can keep this here.